### PR TITLE
search username in namespace annotations

### PR DESCRIPTION
**Pull request by sparkoo**
Signed-off-by: Michal Vala <mvala@redhat.com>

### What does this PR do?
Evaluates username and try to find it in annotations, instead of labels.

TODO:
 - [x] Tests
 - [x] Documentation

### What issues does this PR fix or reference?

https://github.com/eclipse/che/issues/18176

### Documentation PR
https://github.com/eclipse/che-docs/pull/1659

### How to test this PR?

1. Have user with non-DNS username (like an email)
1. Prepare namespace with labels and annotations like and create admin rolebinding for the user in that namespace:
```
metadata:
  annotations:
    che.eclipse.org/username: tester@tester.cz
  labels:
    app.kubernetes.io/component: workspace
    app.kubernetes.io/part-of: che.eclipse.org
```
3. New workspace should be created there, either with dashboard or factory

This also fixes the issue with namespace listing, which was not working at all.
1. Have user with non-DNS username (like an email)
1. Create new custom workspace from dashboard. It should work.

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [x] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [x] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [x] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [x] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [x] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [x] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [x] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [x] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

**Review by skabashnyuk — COMMENTED** [review]
looks good to me. Are you planning to change anything in documentation?

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java [hunk 1]
```diff
@@ -629,20 +640,36 @@ private String evalDefaultNamespace(NamespaceResolutionContext resolutionCtx)
...
    * @return evaluated labels
    */
-  protected Map<String, String> evaluateLabelsPlaceholders(
+  protected Map<String, String> evaluateAnnotationPlaceholders(
       NamespaceResolutionContext namespaceCtx) {
-    Map<String, String> evaluatedLabels = new HashMap<>();
-    for (String labelName : namespaceLabels.keySet()) {
+    Map<String, String> evaluatedAnnotations = new HashMap<>();
+    for (String labelName : namespaceAnnotations.keySet()) {
       String evaluatedLabelValue =
```
Comment (metlos) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java]: `evaluatedAnnotationValue` or `evaluatedValue`?

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java [hunk 2]
```diff
@@ -629,20 +640,36 @@ private String evalDefaultNamespace(NamespaceResolutionContext resolutionCtx)
...
    *
    * @return evaluated labels
    */
-  protected Map<String, String> evaluateLabelsPlaceholders(
+  protected Map<String, String> evaluateAnnotationPlaceholders(
       NamespaceResolutionContext namespaceCtx) {
-    Map<String, String> evaluatedLabels = new HashMap<>();
-    for (String labelName : namespaceLabels.keySet()) {
+    Map<String, String> evaluatedAnnotations = new HashMap<>();
+    for (String labelName : namespaceAnnotations.keySet()) {
```
Comment (metlos) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java]: `annotationName`?

**Review by metlos — APPROVED** [review]
LGTM apart from the below nitpicks.

**Comment by sparkoo** [comment]
> looks good to me. Are you planning to change anything in documentation?

yes https://github.com/eclipse/che-docs/pull/1659

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java (hunk 1 above)
Comment (sparkoo) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java]: yes

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java (hunk 2 above)
Comment (sparkoo) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java]: fixed

**Comment by sparkoo** [comment]
[crw-ci-test]

**Review by skabashnyuk — APPROVED** [review]

**Comment by sparkoo** [comment]
[crw-ci-test]

**Comment by sparkoo** [comment]
[crw-ci-test]

**Comment by sparkoo** [comment]
[crw-ci-test]

**Comment by sparkoo** [comment]
[crw-ci-test]

**Review by mshaposhnik — APPROVED** [review]

**Comment by mshaposhnik** [comment]
Go-go. I need to build my fix ontop of yours :)

**Pull request merged**